sklearn/utils/validation.py

```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+        ...
+    ValueError: All eigenvalues are negative (maximum is -1.000000). The matrix
+        is maybe not PSD, or something went wrong with the eigenvalues
+        decomposition.
+    >>> _check_psd_eigenvalues([5, -1])     # significant negative
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5, 0])
+    >>> _check_psd_eigenvalues([5, -5e-5])  # insignificant negative
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+    >>> _check_psd_eigenvalues([5, 4e-12])  # bad conditioning (too small)
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+
+    """
+
+    lambdas = np.array(lambdas)
+    is_double_precision = lambdas.dtype == np.float64
+
+    # note: the minimum value available is
+    #  - single-precision: np.finfo('float32').eps = 1.2e-07
+    #  - double-precision: np.finfo('float64').eps = 2.2e-16
+
+    # the various thresholds used for validation
+    # we may wish to change the value according to precision.
+    significant_imag_ratio = 1e-5
+    significant_neg_ratio = 1e-5 if is_double_precision else 5e-3
+    significant_neg_value = 1e-10 if is_double_precision else 1e-6
+    small_pos_ratio = 1e-12
+
+    # Check that there are no significant imaginary parts
+    if not np.isreal(lambdas).all():
+        max_imag_abs = np.abs(np.imag(lambdas)).max()
+        max_real_abs = np.abs(np.real(lambdas)).max()
+        if max_imag_abs > significant_imag_ratio * max_real_abs:
+            raise ValueError(
+                "There are significant imaginary parts in eigenvalues (%f "
+                "of the maximum real part). The matrix is maybe not PSD, or "
+                "something went wrong with the eigenvalues decomposition."
+                % (max_imag_abs / max_real_abs))
+
+    # Remove the insignificant imaginary parts
+    lambdas = np.real(lambdas)
+
+    # Check that there are no significant negative eigenvalues
+    max_eig = lambdas.max()
+    if max_eig < 0:
+        raise ValueError("All eigenvalues are negative (maximum is %f). The "
+                         "matrix is maybe not PSD, or something went wrong "
+                         "with the eigenvalues decomposition." % max_eig)
+
+    else:
+        min_eig = lambdas.min()
+        if (min_eig < -significant_neg_ratio * max(max_eig, 0)
+                and min_eig < -significant_neg_value):
+            warnings.warn("There are significant negative eigenvalues "
+                          "(%f of the maximum positive). The matrix is maybe "
+                          "not PSD, or something went wrong with the "
+                          "eigenvalues decomposition. Replacing them with "
+                          "zero." % (-min_eig / max_eig),
+                          PositiveSpectrumWarning)
+
+    # Remove all negative values in all cases
+    lambdas[lambdas < 0] = 0
+
+    # Finally check for conditioning
+    too_small_lambdas = (0 < lambdas) & (lambdas < small_pos_ratio * max_eig)
```

Comment:
with the current logic, `max_eig` could be negative.

Reply:
No, we raise a ValueError above if `max_eig < 0`